File: order-service-vt/src/main/java/com/baeldung/vt/domain/Order.java
package com.baeldung.vt.domain;

import java.util.Date;
import java.util.List;

import org.bson.types.ObjectId;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import com.baeldung.vt.constants.OrderStatus;
import com.baeldung.vt.serdeser.ObjectIdSerializer;
import com.baeldung.vt.serdeser.ObjectIdValueSerializer;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import lombok.Data;

@Data
@Document
public class Order {

    @Id
    // Spring Kafka's JSON (de)serializer still runs on Jackson 2, while Spring MVC runs on Jackson 3 - both annotations are needed.
    @JsonSerialize(using = ObjectIdSerializer.class)
    @tools.jackson.databind.annotation.JsonSerialize(using = ObjectIdValueSerializer.class)
    private ObjectId id;
    private String userId;
    private List<LineItem> lineItems;
    private Long total;
    private String paymentMode;
    private Address shippingAddress;
    private Date shippingDate;
    private OrderStatus orderStatus;
    private String responseMessage;

    public Order setLineItems(List<LineItem> lineItems) {
        this.lineItems = lineItems;
        return this;
    }

    public Order setShippingDate(Date shippingDate) {
        this.shippingDate = shippingDate;
        return this;
    }

    public Order setOrderStatus(OrderStatus orderStatus) {
        this.orderStatus = orderStatus;
        return this;
    }

    public Order setResponseMessage(String responseMessage) {
        this.responseMessage = responseMessage;
        return this;
    }

    // SEC-004: Lombok's @Data toString() recursed into shippingAddress
    // (name/house/street/city/zip) and userId, so every log.info(..., order)
    // call across the saga wrote unredacted customer PII to plaintext logs.
    // Override toString() to log only non-PII fields Lombok would otherwise
    // include automatically.
    @Override
    public String toString() {
        return "Order[id=" + id + ", orderStatus=" + orderStatus
            + ", lineItems=" + (lineItems == null ? 0 : lineItems.size()) + " items]";
    }

}

File: order-service-cs/Domain/Order.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using OrderService.Api.Constants;

namespace OrderService.Api.Domain;

/// <summary>Mirrors com.baeldung.domain.Order in order-service (Java) field-for-field.</summary>
public class Order
{
    // MongoDB.Driver's default ID convention maps a member named "Id"
    // (case-insensitively) to the document's "_id" field, matching Spring
    // Data's @Id -> _id mapping in the Java reference.
    public ObjectId Id { get; set; }

    public string? UserId { get; set; }

    public List<LineItem>? LineItems { get; set; }

    public long? Total { get; set; }

    public string? PaymentMode { get; set; }

    public Address? ShippingAddress { get; set; }

    public DateTime? ShippingDate { get; set; }

    // Stored as its member name string in MongoDB (matching Spring Data's
    // default enum-as-name-string mapping), not the driver's default
    // integer representation.
    [BsonRepresentation(BsonType.String)]
    public OrderStatus OrderStatus { get; set; }

    public string? ResponseMessage { get; set; }

    // SEC-004: a default-generated ToString() would recurse into
    // ShippingAddress (name/house/street/city/zip) and UserId, writing
    // unredacted customer PII to plaintext logs on every log call that
    // interpolates an Order. Override ToString() to log only non-PII fields,
    // matching order-service (Java)'s override.
    public override string ToString() =>
        $"Order[id={Id}, orderStatus={OrderStatus}, lineItems={LineItems?.Count ?? 0} items]";
}
